File: app/src/main/java/com/mintthursday/IngredientEditAdapter.java
package com.mintthursday;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

public class IngredientEditAdapter extends RecyclerView.Adapter<IngredientEditAdapter.IngredientEditViewHolder> {
    OnItemClickListenerIngredient ingredientListener;
    private List<Ingredient> ingredientList = new ArrayList<>();

    public void setOnItemClickListener(final OnItemClickListenerIngredient itemClickListener) {
        this.ingredientListener = itemClickListener;
    }

    public List<Ingredient> getItems() {
        return ingredientList;
    }

    public void setItems(List<Ingredient> ingredients) {
        ingredientList = ingredients;
        notifyDataSetChanged();
    }

    public void addItem(Ingredient ingredient) {
        ingredientList.add(ingredient);
        notifyItemInserted(ingredientList.indexOf(ingredient));
    }

    private void deleteItem(int position) {
        ingredientList.remove(position);
        notifyItemRemoved(position);
    }

    public void updateItem(Ingredient ingredient, int itemPosition) {
        ingredientList.set(itemPosition, ingredient);
        notifyItemChanged(itemPosition);
    }

    @NonNull
    @NotNull
    @Override
    public IngredientEditAdapter.IngredientEditViewHolder onCreateViewHolder(@NonNull @NotNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.view_ingredient_edit, parent, false);
        return new IngredientEditViewHolder(view);
    }

    @Override
    public void onBindViewHolder(IngredientEditViewHolder holder, int position) {
        holder.bind(ingredientList.get(position));

    }

    @Override
    public int getItemCount() {
        return ingredientList.size();
    }

    class IngredientEditViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        private TextView ingrName;
        private TextView ingrQty;

        private TextView ingrUnit;

        private ImageButton btnIngrRemove;


        public IngredientEditViewHolder(@NonNull @NotNull View itemView) {
            super(itemView);
            ingrName = itemView.findViewById(R.id.ingr_name);
            ingrQty = itemView.findViewById(R.id.ingr_qty);
            ingrUnit = itemView.findViewById(R.id.ingr_unit);
            btnIngrRemove = itemView.findViewById(R.id.ingr_remove);
            btnIngrRemove.setOnClickListener(this);
            itemView.setOnClickListener(this);
        }

        public void bind(Ingredient ingredient) {
            ingrName.setText(ingredient.getName());
            ingrQty.setText(String.valueOf(ingredient.getQuantity()));
            ingrUnit.setText(ingredient.getUnit());
        }

        @Override
        public void onClick(View v) {
            if (v.equals(btnIngrRemove)) {
                deleteItem(getAdapterPosition());
            } else {
                if (ingredientListener != null) {
                    Ingredient ingredient = ingredientList.get(getAdapterPosition());
                    int i = getAdapterPosition();
                    ingredientListener.onItemClick(ingredient, i);
                }
            }
        }

    }


}

File: app/src/main/java/com/mintthursday/recipe/creation/ingredientcreation/IngredientEditAdapter.kt
package com.mintthursday.recipe.creation.ingredientcreation

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.mintthursday.R
import com.mintthursday.models.Ingredient
import com.mintthursday.recipe.creation.ingredientcreation.IngredientEditAdapter.IngredientEditViewHolder

class IngredientEditAdapter : RecyclerView.Adapter<IngredientEditViewHolder>() {

    private lateinit var ingredientListener: OnIngredientClickListener
    var items = mutableListOf<Ingredient>()
        private set

    fun setItems(items: List<Ingredient>) {
        this.items = items.toMutableList()
        notifyItemRangeInserted(0, items.size)
    }

    fun setOnItemClickListener(itemClickListener: OnIngredientClickListener) {
        ingredientListener = itemClickListener
    }

    fun addItem(ingredient: Ingredient) {
        items.add(ingredient)
        notifyItemInserted(items.indexOf(ingredient))
    }

    private fun deleteItem(position: Int) {
        items.removeAt(position)
        notifyItemRemoved(position)
    }

    fun updateItem(ingredient: Ingredient, itemPosition: Int) {
        items[itemPosition] = ingredient
        notifyItemChanged(itemPosition)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): IngredientEditViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(
                R.layout.view_ingredient_edit,
                parent,
                false
            )
        return IngredientEditViewHolder(view)
    }

    override fun onBindViewHolder(holder: IngredientEditViewHolder, position: Int) {
        holder.bind(items[position])
    }

    override fun getItemCount(): Int {
        return items.size
    }

    inner class IngredientEditViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {

        private val ingrName: TextView = itemView.findViewById(R.id.ingr_name)
        private val ingrQty: TextView = itemView.findViewById(R.id.ingr_qty)
        private val ingrUnit: TextView = itemView.findViewById(R.id.ingr_unit)
        private val btnIngrRemove: ImageButton = itemView.findViewById(R.id.ingr_remove)

        fun bind(ingredient: Ingredient) {
            ingrName.text = ingredient.name
            ingrQty.text = ingredient.quantity.toString()
            ingrUnit.text = ingredient.unit
        }

        override fun onClick(v: View) {
            if (v == btnIngrRemove) {
                deleteItem(adapterPosition)
            } else {
                val ingredient = items[adapterPosition]
                ingredientListener.onItemClick(ingredient, adapterPosition)
            }
        }

        init {
            btnIngrRemove.setOnClickListener(this)
            itemView.setOnClickListener(this)
        }
    }

    interface OnIngredientClickListener {
        fun onItemClick(ingredient: Ingredient, itemPosition: Int)
    }
}